src/Compilers/CSharp/Portable/Symbols/DelegateCaches/TypeParameterUsageChecker.cs

```diff
@@ -0,0 +1,163 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Immutable;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.CSharp.Symbols
+{
+    using TypeParams = ImmutableArray<TypeParameterSymbol>;
+
+    /// <summary>
+    /// Checks whether a type or a method is constructed with some type that involves the specified generic type parameters.
+    /// </summary>
+    internal class TypeParameterUsageChecker : CSharpSymbolVisitor<TypeParams, bool>
+    {
+        public static TypeParameterUsageChecker Instance { get; } = new TypeParameterUsageChecker();
+
+        public override bool VisitTypeParameter(TypeParameterSymbol symbol, TypeParams argument) => argument.Contains(symbol);
+
+        public override bool VisitArrayType(ArrayTypeSymbol symbol, TypeParams argument)
+        {
+            if (Visit(symbol.ElementType, argument))
+            {
+                return true;
+            }
+
+            if (VisitCustomModifiers(symbol.CustomModifiers, argument))
+            {
+                return true;
+            }
+
+            return false;
+        }
+
+        public override bool VisitNamedType(NamedTypeSymbol symbol, TypeParams argument)
+        {
+            if (symbol.IsAnonymousType)
+            {
+                foreach (var anonPropType in AnonymousTypeManager.GetAnonymousTypePropertyTypes(symbol))
+                {
+                    if (Visit(anonPropType, argument))
+                    {
+                        return true;
+                    }
+                }
+
+                return false;
+            }
+
+            foreach (var typeArg in symbol.TypeArguments)
+            {
+                if (Visit(typeArg, argument))
+                {
+                    return true;
+                }
+            }
+
+            if (symbol.HasTypeArgumentsCustomModifiers)
+            {
+                foreach (var customModifiers in symbol.TypeArgumentsCustomModifiers)
+                {
+                    if (VisitCustomModifiers(customModifiers, argument))
+                    {
+                        return true;
+                    }
+                }
+
+                return false;
```

Comment:
Ops, how could that happen.

PS. I find that we have existing extension methods could potentially replace my two visitors. See [these two overloads](https://github.com/dotnet/roslyn/blob/master/src/Compilers/CSharp/Portable/Symbols/TypeSymbolExtensions.cs#L703-L724), we can make another 2 for MethodSymbols on top of  them. But VisitType does not visit custom modifiers, can we add?